README.md
# e101-server
The bluetooth server used for motor control in the E101 006 Group 4 Automatic Chicken Coop Door Opener Project. Requests are sent from an accompanying android app. 
- Actuation may be controled in real time using OPEN and CLOSE functions
- It may also be scheduled in advance for a specific day and time
- Motor control uses gpiozero
- Tested only on a raspberrypi 4b using vex motors


history.txt
    1  ls
    2  pwd
    3  sudo apt update
    4  apt search bluez
    5  sudo apt install python3-bluez
    6  python
    7  ls
    8  python server.py 
    9  pip install httplib2
   10  apt search pip
   11  sudo apt install python3-pip
   12  python server.py 
   13  apt search httplib2
   14  sudo apt install python3-httplib2
   15  python server.py 
   16  vi server.py 
   17  python server.py 
   18  vi server.py 
   19  sudo python server.py
   20  vi server.py 
   21  sudo python server.py
   22  sudo nano /etc/systemd/system/dbus-org.bluez.service 
   23  sudo systemctl start bluetooth
   24  bluetoothctl 
   25  sudo python server.py
   26  hciconfig hciX piscan
   27  sudo hciconfig hciX piscan
   28  sudo python server.py
   29  python server.py
   30  sudo nano /etc/systemd/system/dbus-org.bluez.service 
   31  bluetoothd
   32  sudo bluetoothd
   33  sudo nano /etc/systemd/system/dbus-org.bluez.service 
   34  sudo systemctl restart bluetooth
   35  systemctl daemon-reload
   36  python server.py
   37  sudo sdptool add SP
   38  python server.py
   39  sudo python server.py
   40  sudo systemctl restart bluetooth
   41  python server.py
   42  sudo systemctl status bluetooth
   43  less /var/run/sdp 
   44  sudo sdptool add SP
   45  python server.py
   46  sudo python server.py
   47  rfcomm
   48  sudo nano /etc/systemd/system/dbus-org.bluez.service 
   49  reboot
   50  sudo reboot
   51  sudo python server.py
   52  systemctl status bluetooth
   53  sudo sdptool add SP
   54  systemctl status bluetooth
   55  sudo python server.py
   56  bluetoothctl 
   57  sudo python server.py
   58  bluetoothctl 
   59  sudo python server.py
   60  ls
   61  sudo apt install neovim
   62  history 
   63  history >> history.txt


test.py
import data
from datetime import datetime

d = data.Data(Command="Open", Time=datetime.now().isoformat(), T1="", T2="")
print(d.Serialize())

d = data.Data(Command="Sched", T1=datetime.now().isoformat(), T2="TIME")
print(d.Serialize())


motor.py
# Author: Ryan Clifford
# 2022-03-26
import gpiozero
import os
import time


class MotorController:
    OPEN_VALUE = 1
    CLOSE_VALUE = -1
    RUN_TIME = 3
    current_status = ""

    def __init__(
        self,
        currentState,
        gpio_pin=17,
        min_pulse_width=0.001,
        max_pulse_width=0.002,
        frame_width=0.020,
    ) -> None:
        os.system("sudo pigpiod")
        gpiozero.Device.pin_factory = gpiozero.pins.pigpio.PiGPIOFactory()
        self.motor = gpiozero.Servo(
            pin=gpio_pin,
            initial_value=0,
            min_pulse_width=min_pulse_width,
            max_pulse_width=max_pulse_width,
            frame_width=frame_width,
        )
        self.current_status = currentState

    def Pin(self):
        return self.motor.pin

    def Value(self):
        return self.motor.value

    def __set_value(self, value):
        self.motor.value = value

    def stop(self):
        self.__set_value(0)

    def open(self):
        if self.current_status != "open":
            self.__set_value(self.OPEN_VALUE)
            time.sleep(self.RUN_TIME)
            self.stop()
            self.current_status = "open"

    def close(self):
        if self.current_status != "closed":
            self.__set_value(self.CLOSE_VALUE)
            time.sleep(self.RUN_TIME)
            self.stop()
            self.current_status = "closed"

    def schedule(self, T1, T2):
        raise NotImplementedError


if __name__ == "__main__":
    print("Try server.py")


schedule.py
import datetime
import logging
from multiprocessing import Lock
from threading import Thread
from time import sleep
import motor


class Scheduler:
    openTime: datetime.datetime = datetime.datetime.now()
    closeTime: datetime.datetime = openTime
    motorController: motor.MotorController

    def __init__(
        self, motorController, motorLock, date, openTime, closeTime, currentState
    ) -> None:
        try:
            date = DateParser(date)
        except:
            logging.critical("Could not parse date")
        try:
            self.openTime = TimeParser(openTime, date)
            self.closeTime = TimeParser(closeTime, date)
        except:
            logging.critical("Could not parse time")
        self.currentState = currentState
        self.motorController = motorController
        self.motorLock = motorLock

    def start(self):
        return Thread(target=self.actuate)

    def actuate(self):
        waitTime = 0  # time before actuation in seconds
        nextWait = 0
        actuator = None
        logging.info(f"Open: {self.openTime}")
        logging.info(f"Close: {self.closeTime}")
        openWait = (self.openTime - datetime.datetime.now()).total_seconds()
        closeWait = (self.closeTime - datetime.datetime.now()).total_seconds()
        if openWait < closeWait:
            waitTime = openWait
            nextWait = closeWait - waitTime
            logging.info(f"Opening in {waitTime} seconds")
            actuator = self.motorController.open
            nextActuator = self.motorController.close
        else:
            waitTime = closeWait
            nextWait = openWait - waitTime
            logging.info(f"Closing in {waitTime} seconds")
            actuator = self.motorController.close
            nextActuator = self.motorController.open
        sleep(waitTime)
        with self.motorLock:
            logging.info("Opening/Closing")
            actuator()
        logging.info(f"Sleeping {nextWait} for next actuation")
        sleep(nextWait)
        with self.motorLock:
            logging.info("Opening/Closing")
            nextActuator()


# Non standard date time format given as strings
def DateParser(date: str) -> datetime.datetime:
    logging.debug(f"DateParser: {date}")
    parts = date.split("-")
    year = int(parts[0].strip())
    month = int(parts[1].strip())
    day = int(parts[2].strip())
    return datetime.datetime(year, month, day)


def TimeParser(time: str, date: datetime.datetime) -> datetime.datetime:
    logging.debug(f"TimeParser: {time}")
    parts = time.split(":")
    hour = int(parts[0].strip())
    minute = int(parts[1].strip())
    computed_time = datetime.time(hour, minute, 0)
    return datetime.datetime.combine(date.date(), computed_time)


server.py
#!/usr/bin/env python3

# based off: https://github.com/pybluez/pybluez/blob/master/examples/simple/rfcomm-server.py
# Author: Ryan Clifford
# 2022-03-26

from datetime import datetime
import sys
import bluetooth
import motor
import logging
import schedule
from threading import Thread, Lock

logging.basicConfig(
    filename="server.log",
    level=logging.INFO,
    format="[%(asctime)s] %(levelname)s: %(message)s",
)
if len(sys.argv) < 2:
    logging.critical("Current door state not provided")
    raise Exception('Provide current state of door ("open"/"closed")')
DEBUG_MODE = False
if len(sys.argv) > 2:
    if sys.argv[2].lower() == "debug":
        DEBUG_MODE = True
        print("In debugging mode, will not call motor control")
        logging.info("Started in Debug Mode")

SERVER_NAME = "Group 4"
logging.info(f"Server name: {SERVER_NAME}")
UUID = "9f32d32e-e7b2-484b-819f-a571b8219a74"
logging.info(f"Service UUID: {UUID}")
BUFSIZE = 4096
logging.info(f"RECV BUFFSIZE: {BUFSIZE}")
socket = bluetooth.BluetoothSocket(bluetooth.RFCOMM)
socket.bind(("", bluetooth.PORT_ANY))
socket.listen(1)

port = socket.getsockname()[1]

bluetooth.advertise_service(
    socket,
    SERVER_NAME,
    service_id=UUID,
    service_classes=[UUID, bluetooth.SERIAL_PORT_CLASS],
    profiles=[bluetooth.SERIAL_PORT_PROFILE],
)
logging.info("Bluetooth Service Started")
currentState = sys.argv[1].lower()
motorLock = Lock()
if not DEBUG_MODE:
    MotorControl = motor.MotorController(currentState=currentState)
    logging.info("Motor Controller Initialized")


currentTime = datetime.now()
try:
    while True:
        client_sock, client_info = socket.accept()
        logging.info(f"Connect to: {client_info}")

        try:
            while True:
                # assume all data can be recieved at once
                data = client_sock.recv(BUFSIZE)
                if not data:
                    break
                logging.debug(f"RECV: {data}")

                # data = json.loads(data)
                if data == b"OPEN":
                    if currentState == "closed":
                        with motorLock:
                            logging.info(f"[{data}] Opening...")
                            if not DEBUG_MODE:
                                MotorControl.open()
                            currentState = "open"
                    else:
                        logging.info("Door already open")
                elif data == b"CLOSE":
                    if currentState == "open":
                        with motorLock:
                            logging.info(f"[{data}] Closing...")
                            if not DEBUG_MODE:
                                MotorControl.close()
                                currentState = "closed"
                            else:
                                logging.info("Door already closed")
                elif data.decode().split(";")[0] == "SCHED":
                    #logging.error(f"Scheduling not implemented: [{data}]")
                    schedule_input = data.decode().split(";")
                    logging.debug(f"SCHED Case: {schedule_input}")
                    logging.info(schedule_input)
                    if len(schedule_input) < 4:
                        logging.error(f"Schedule called with data: {data}")
                    scheduler = schedule.Scheduler(
                        MotorControl,
                        motorLock,
                        schedule_input[1],
                        schedule_input[2],
                        schedule_input[3],
                        currentState,
                    )
                    scheduler.start().start()
                    logging.info(f"Started scheduler")
                    # raise NotImplementedError
                else:
                    logging.info(f"Unknown data: {data}")
        except OSError:
            pass

        logging.info("Client Disconnected")

        client_sock.close()
except KeyboardInterrupt:
    socket.close()
    logging.info("\nSocket Closed")
